core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -0,0 +1,209 @@
+            topicPartitionData = acquirablePartitions();
+        // tryComplete invoked forceComplete, so we can use the data from tryComplete.
+        else
+            topicPartitionData = topicPartitionDataFromTryComplete;
+        try {
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request with an empty response.
+                shareFetchPartitionData.future().complete(Collections.emptyMap());
+                return;
+            }
+            log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
+                    topicPartitionData, shareFetchPartitionData.groupId(), shareFetchPartitionData.fetchParams());
+
+            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
+                shareFetchPartitionData.fetchParams(),
+                CollectionConverters.asScala(
+                    topicPartitionData.entrySet().stream().map(entry ->
+                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
+                ),
+                QuotaFactory.UnboundedQuota$.MODULE$,
+                true);
+
+            List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = new ArrayList<>();
+            responseLogResult.foreach(tpLogResult -> {
+                TopicIdPartition topicIdPartition = tpLogResult._1();
+                LogReadResult logResult = tpLogResult._2();
+                FetchPartitionData fetchPartitionData = logResult.toFetchPartitionData(false);
+                responseData.add(new Tuple2<>(topicIdPartition, fetchPartitionData));
+                return BoxedUnit.UNIT;
+            });
+
+            log.trace("Data successfully retrieved by replica manager: {}", responseData);
+            ShareFetchUtils.processFetchResponse(shareFetchPartitionData, responseData, partitionCacheMap, replicaManager)
+                .whenComplete((result, throwable) -> {
+                    if (throwable != null) {
+                        log.error("Error processing fetch response for share partitions", throwable);
+                        shareFetchPartitionData.future().completeExceptionally(throwable);
+                    } else {
+                        shareFetchPartitionData.future().complete(result);
+                    }
+                    // Releasing the lock to move ahead with the next request in queue.
+                    releasePartitionLocks(shareFetchPartitionData.groupId(), topicPartitionData.keySet());
+                    // If we have a fetch request completed for a topic-partition, it means  the HWM has advanced,
+                    // then we should check if there is a pending share fetch request for the topic-partition and complete it.
+                    result.keySet().forEach(topicIdPartition -> delayedShareFetchPurgatory.checkAndComplete(
```

Comment:
Hi @junrao, yes, it makes total sense. I have removed this part of code in my latest commits. I have created a JIRA https://issues.apache.org/jira/browse/KAFKA-17509 to add a delayed actions queue for this purpose and will address it in my future PRs, since this PR has become quite big already.